Mathlib/RingTheory/TensorProduct/MonoidAlgebra.lean

```diff
@@ -0,0 +1,420 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.LinearAlgebra.DirectSum.Finsupp
+import Mathlib.RingTheory.TensorProduct
+import Mathlib.Algebra.Algebra.Equiv
+
+/-! # TensorProduct of a monoid algebra
+
+This file consists of three sections, the first two will have to be moved
+somewhere else.
+
+## two lemmas about `Finsupp.single`
+
+## functoriality for `MonoidAlgebra`
+
+We prove some functoriality definitions for the `Semiring` variable.
+It remains to prove that this is a functor.
+
+## tensor product
+
+* `MonoidAlgebra.rTensorAlgEquiv`, the tensor product of
+   `MonoidAlgebra M α` with `N` is `R`-linearly equivalent to `MonoidAlgebra (M ⊗[R] N) α`
+
+* `MonoidAlgebra.scalarRTensorAlgEquiv`, the tensor product of
+   `MonoidAlgebra R α` with `N` is `R`-linearly equivalent to `MonoidAlgebra N α`
+
+## TODO
+
+* State and prove functoriality properties
+
+* Prove the `lTensor` variants
+
+* Combine both to get that the tensor product of `MonoidAlgebra M α`
+  with `MonoidAlgebra N β` is `R`-linearly equivalent to
+  `MonoidAlgebra (M ⊗[R] N) (α × β)`.
+-/
+
+open TensorProduct
+
+variable {α : Type*} [Monoid α] [DecidableEq α]
+  {R : Type*} [CommSemiring R]
+  {M N P : Type*}
+
+
+section Finsupp
+
+namespace Finsupp
+
+lemma apply_single [AddCommMonoid N] [AddCommMonoid P]
+    (e : N →+ P) (a : α) (n : N) (b : α) :
+    e ((single a n) b) = single a (e n) b := by
+  simp only [single_apply]
+  split_ifs; rfl; exact map_zero e
+
+lemma mapRange.addMonoidHom_apply_single
+    [AddCommMonoid N] [AddCommMonoid P] (e : N →+ P) (a : α) (n : N) :
+    mapRange.addMonoidHom e (single a n) = single a (e n) := by
+    simp only [addMonoidHom_apply, mapRange_single]
+--    ext b
+--    exact apply_single e a n b
+
+end Finsupp
+
+end Finsupp
+
+section Functoriality
+
+namespace MonoidAlgebra
+
+open Finsupp
+
+variable [Semiring N] [Semiring P]
+
+/- noncomputable example
+    {P : Type*} [Semiring P] [Algebra R P] (e : N ≃ₗ[R] P) :
+    MonoidAlgebra N α ≃ₗ[R] MonoidAlgebra P α :=
+  Finsupp.mapRange.linearEquiv e -/
+
+lemma apply_single
+    (e : N →+ P) (a : α) (n : N) (b : α) :
+    e ((single a n) b) = single a (e n) b :=
+  Finsupp.apply_single e a n b
+
+
+/-- RingHom functoriality for the monoid algebra -/
+noncomputable def ringHom (e : N →+* P) :
+    MonoidAlgebra N α →+* MonoidAlgebra P α := {
```

Comment:
The AddMonoidAlgebra version was done [here](https://github.com/leanprover-community/mathlib4/pull/7219/files#diff-ef701cddec9d543cfb4ea50b0bf2519305cc560a464990b50b13dca605614a1cR19-R20) in #7219 with more general typeclass assumptions. I think `RingHom.mapMonoidAlgebra` is probably a better name, as it allows dot notations. I don't agree with the name `ringHomCongrLeft` in #7219 since `congr` should be reserved for equalities or Equiv.

Reply:
I generalized to allow `MulOneClass`.
Regarding the names, I don't feel responsible for somebody elses' PR but will obey suggestions.